### ISSUE TITLE
Using zero optimizer with torch.optim.lr_scheduler will hit error: Zero is not an Optimizer

### ISSUE THREAD

**Issue by LiweiPeng** [microsoft/DeepSpeed#122]
torch.optim.lr_scheduler provides base class for lr scheduler design. E.g. NVIDIA BERT squad code uses it to define its lr schedulers (https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/run_squad.py#L1019).

However, using the above scheduler with zero optimizer like below will hit exception 'Zero is not an optimizer'. This is because Zero optimizer class is not a subclass of torch.optim.Optimizer.

model, optimizer, _, _ = deepspeed.initialize(...)
scheduler = LinearWarmUpScheduler(optimizer, ...)

While this issue can be worked around like deepspeed example code here (https://github.com/microsoft/DeepSpeedExamples/blob/master/BingBertSquad/nvidia_run_squad_deepspeed.py#L919), it will be better to be able to use LinearWarmUpScheduler (https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/schedulers.py#L87) class directly.

Can you tell me how to use LinearWarmUpScheduler directly? Thanks.

**Comment by tjruwase** [comment]
It appears the most straight forward solution here is to make Zero is a subclass of torch.optim.Optimizer. We will work on this. Please stay tuned.

**Comment by tjruwase** [comment]
I don't think there is a straight forward solution without modifying LinearWarmUpScheduler() to remove this assert. Be mindful, however, that there is a probably a good reason for the assert in the first place. Since LinearWarmUpScheduler is not DeepSpeed code, you will need to do the modification yourself. We are certainly interested in learning about your experience, so please share.

**Comment by LiweiPeng** [comment]
Thanks @tjruwase for the quick update. This torch.optim.Optimizer assert is not only in LinearWarmUpScheduler, but also in torch.optim.lr_scheduler._LRScheduler. So simply modifying LinearWarmUpScheduler won't work.

DeepSpeed implements several lr schedulers. Can you add this LinearWarmUpScheduler used by BERT?

**Comment by tjruwase** [comment]
I see. How about this?
1) Can you see if [WarmupLR](https://github.com/microsoft/DeepSpeed/blob/5f6294bd04cbf099c2dc15b9aebc44ebe91f0044/deepspeed/pt/deepspeed_lr_schedules.py#L582) can meet your requirements?
2) If not, can you please contribute LinearWarmUpScheduler to DeepSpeed?

Thanks

**Comment by LiweiPeng** [comment]
WarmupLR is different from the [LinearWarmUpScheduler](https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/schedulers.py#L87).  WarmupLR doesn't have decay. LinearWarmUpScheduler has a warmup phase and decay phase.

I am using a work-around now.  I may create a PR in the future.

**Comment by Colanim** [comment]
Do you mind sharing your workaround ? I'm also having this issue

**Comment by LiweiPeng** [comment]
I used [deepspeed example code](https://github.com/microsoft/DeepSpeedExamples/blob/master/BingBertSquad/nvidia_run_squad_deepspeed.py#L919)

**Comment by tjruwase** [comment]
Closing since there is a workaround.
